stellargraph/layer/watch_your_step.py

```diff
@@ -0,0 +1,199 @@
+class WatchYourStep:
+    """
+    Implementation of the node embeddings as in Watch Your Step: Learning Node Embeddings via Graph Attention
+    https://arxiv.org/pdf/1710.09599.pdf.
+
+    This model requires specification of the number of random walks starting from each node, and the embedding dimension
+    to use for the node embeddings. Note, that the embedding dimension should be an even number or else will be
+    rounded down to nearest even number.
+
+    .. warning::
+
+        This class is experimental: it is insufficiently tested.
+
+    Args:
+        generator (AdjacencyPowerGenerator): the generator
+        num_walks (int): the number of random walks starting at each node to use when calculating the expected random
+        walks.
+        embedding dimension (int): the dimension to use for the node embeddings
+    """
+
+    def __init__(
+        self, generator, num_walks, embedding_dimension, attention_regularizer=None
+    ):
+
+        if not isinstance(generator, AdjacencyPowerGenerator):
+            raise TypeError(
+                "generator should be an instance of AdjacencyPowerGenerator."
+            )
+
+        if not isinstance(num_walks, int):
+            raise TypeError("num_walks should be an int.")
+
+        if num_walks <= 0:
+            raise ValueError("num_walks should be a positive int.")
+
+        self.num_walks = num_walks
+        self.num_powers = generator.num_powers
+        self.n_nodes = int(generator.Aadj_T.shape[0])
+        self.embedding_dimension = embedding_dimension
+        self.attention_regularizer = attention_regularizer
+
+    def build(self):
+        """
+        This function builds the layers for a keras model.
+
+        returns:
+            A tuple of (inputs, outputs) to use with a keras model.
+        """
+
+        input_rows = Input(batch_shape=(None,), name="row_node_ids", dtype="int64")
+        input_powers = Input(batch_shape=(None, self.num_powers, self.n_nodes))
+
+        left_embedding = Embedding(
+            self.n_nodes,
+            self.embedding_dimension,
+            input_length=None,
+            name="LEFT_EMBEDDINGS",
+        )
+
+        right_embedding = Embedding(
+            self.n_nodes,
+            self.embedding_dimension,
+            input_length=None,
+            name="RIGHT_EMBEDDINGS",
+        )
+
+        vectors_left = Lambda(lambda x: K.transpose(x))(left_embedding(input_rows))
+
+        # always use all left vectors - currently wastes time looking up embeddings
+        # to keep keras happy
+        # TODO: replace the embedding layer with a custom layer to avoid lookups
+        # input cols but be somehow connected to the input to keep keras happy
+        input_cols = Lambda(
+            lambda x: tf.constant(np.arange(int(self.n_nodes)), dtype="int64")
+        )(input_rows)
+        vectors_right = right_embedding(input_cols)
+
+        dot_product = Lambda(lambda x: K.transpose(K.dot(x[0], x[1])))(
```

Comment:
Just to confirm: the paper text uses the "outer product" here (end of page 2), but this seems to be using the inner product/normal matrix multiplication. However... it seems like the paper text is wrong (or at least misleading here), since this is operating on L (n &times; d/2) and R (n &times; d/2) to get a matrix compatible with A (n &times; n), and the outer product would give a tensor of shape n &times; d/2 &times; n &times; d/2 unlike the normal matrix multiplication L &times; R<sup>T</sup> in the equations, which gives shape n &times; n as desired.

Reply:
I thought an outer product was just: `outer(A, B) = A \dot B^T` but maybe that only applies to vectors? But the paper had another inaccuracy so who knows :) 

Either way I think the desired behaviour is taking the inner product of every right vector with every left vector as a similarity measure